ydb/core/wrappers/fs_storage.cpp

```diff
@@ -0,0 +1,508 @@
+#include "fs_storage.h"
+
+#include <ydb/core/base/appdata.h>
+#include <ydb/library/actors/core/actor_bootstrapped.h>
+#include <ydb/library/actors/core/hfunc.h>
+
+#include <contrib/libs/aws-sdk-cpp/aws-cpp-sdk-core/include/aws/core/client/AWSError.h>
+#include <contrib/libs/aws-sdk-cpp/aws-cpp-sdk-s3/include/aws/s3/S3Errors.h>
+
+#include <util/folder/dirut.h>
+#include <util/folder/path.h>
+#include <util/stream/file.h>
+#include <util/system/fs.h>
+#include <util/generic/guid.h>
+
+#include <type_traits>
+
+namespace NKikimr::NWrappers::NExternalStorage {
+
+#define FS_LOG_I(stream) LOG_INFO_S(*TlsActivationContext, NKikimrServices::FS_WRAPPER, stream)
+#define FS_LOG_W(stream) LOG_WARN_S(*TlsActivationContext, NKikimrServices::FS_WRAPPER, stream)
+#define FS_LOG_E(stream) LOG_ERROR_S(*TlsActivationContext, NKikimrServices::FS_WRAPPER, stream)
+
+namespace {
+
+class TFsOperationActor : public NActors::TActorBootstrapped<TFsOperationActor> {
+private:
+    const TString BasePath;
+
+    struct TMultipartUploadSession {
+        const TString Key;
+        TFile File;
+        ui64 TotalSize = 0;
+
+        explicit TMultipartUploadSession(const TString& key)
+            : Key(key)
+            , File(key, CreateAlways | WrOnly | ForAppend)
+        {
+            File.Flock(LOCK_EX);
+        }
+    };
+
+    THashMap<TString, std::unique_ptr<TMultipartUploadSession>> ActiveUploads;
+
+    template<typename TEvResponse>
+    struct RequiresKey : std::true_type {};
+
+    template<>
+    struct RequiresKey<TEvListObjectsResponse> : std::false_type {};
+
+    template<>
+    struct RequiresKey<TEvDeleteObjectsResponse> : std::false_type {};
+
+    template<typename TEvResponse>
+    static constexpr bool HasKeyConstructor() {
+        return RequiresKey<TEvResponse>::value;
+    }
+
+    template<typename TEvResponse>
+    void ReplySuccess(const NActors::TActorId& sender, const std::optional<TString>& key) {
+        typename TEvResponse::TAwsResult awsResult;
+        Aws::Utils::Outcome<typename TEvResponse::TAwsResult, Aws::S3::S3Error> outcome(std::move(awsResult));
+
+        std::unique_ptr<TEvResponse> response;
+        if constexpr (HasKeyConstructor<TEvResponse>()) {
+            Y_ENSURE(key, "Key is required for this response type");
+            response = std::make_unique<TEvResponse>(*key, std::move(outcome));
+        } else {
+            response = std::make_unique<TEvResponse>(std::move(outcome));
+        }
+        TlsActivationContext->AsActorContext().Send(sender, response.release());
```

Comment:
Я только сейчас заметил. А почему так сложно? Можно же проще:
```suggestion
        this->Send(sender, response.release());
```

И далее везде по коду.